File: app/backend/java/payment-service/src/main/java/com/srbms/paymentservice/controller/PaymentController.java
package com.srbms.paymentservice.controller;

import com.srbms.paymentservice.dto.*;
import com.srbms.paymentservice.entity.Wallet;
import com.srbms.paymentservice.entity.WalletTransaction;
import com.srbms.paymentservice.service.PaymentService;
import com.srbms.paymentservice.service.WalletService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/payments")
public class PaymentController {

    private final PaymentService paymentService;
    private final WalletService walletService;

    @Autowired
    public PaymentController(PaymentService paymentService, WalletService walletService) {
        this.paymentService = paymentService;
        this.walletService = walletService;
    }

    @GetMapping("/health")
    public ResponseEntity<ApiResponse<String>> healthCheck() {
        return ResponseEntity.ok(ApiResponse.success("Payment Service is UP and running", "OK"));
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<PaymentResponseDTO>>> getAllPayments() {
        List<PaymentResponseDTO> response = paymentService.getAllPayments();
        return ResponseEntity.ok(ApiResponse.success("All payments retrieved", response));
    }

    @GetMapping("/all")
    public ResponseEntity<ApiResponse<List<PaymentResponseDTO>>> getAllPaymentsAlias() {
        List<PaymentResponseDTO> response = paymentService.getAllPayments();
        return ResponseEntity.ok(ApiResponse.success("All payments retrieved", response));
    }

    @PostMapping("/process")
    public ResponseEntity<ApiResponse<PaymentResponseDTO>> processPayment(@Valid @RequestBody PaymentRequestDTO requestDTO) {
        PaymentResponseDTO response = paymentService.processPayment(requestDTO);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success("Payment processed successfully", response));
    }

    @GetMapping("/{paymentId}")
    public ResponseEntity<ApiResponse<PaymentResponseDTO>> getPaymentById(@PathVariable Integer paymentId) {
        PaymentResponseDTO response = paymentService.getPaymentById(paymentId);
        return ResponseEntity.ok(ApiResponse.success("Payment details retrieved", response));
    }

    @GetMapping("/transaction/{transactionId}")
    public ResponseEntity<ApiResponse<PaymentResponseDTO>> getPaymentByTransactionId(@PathVariable String transactionId) {
        PaymentResponseDTO response = paymentService.getPaymentByTransactionId(transactionId);
        return ResponseEntity.ok(ApiResponse.success("Payment details retrieved", response));
    }

    @GetMapping("/ride/{rideId}")
    public ResponseEntity<ApiResponse<PaymentResponseDTO>> getPaymentByRideId(@PathVariable Integer rideId) {
        PaymentResponseDTO response = paymentService.getPaymentByRideId(rideId);
        return ResponseEntity.ok(ApiResponse.success("Payment details retrieved", response));
    }

    @GetMapping("/user/{userId}")
    public ResponseEntity<ApiResponse<List<PaymentResponseDTO>>> getPaymentsByUserId(@PathVariable Integer userId) {
        List<PaymentResponseDTO> response = paymentService.getPaymentsByUserId(userId);
        return ResponseEntity.ok(ApiResponse.success("User payment history retrieved", response));
    }

    @PostMapping("/refund")
    public ResponseEntity<ApiResponse<PaymentResponseDTO>> refundPayment(@Valid @RequestBody RefundRequestDTO requestDTO) {
        PaymentResponseDTO response = paymentService.refundPayment(requestDTO);
        return ResponseEntity.ok(ApiResponse.success("Payment refunded successfully", response));
    }

    @PostMapping("/calculate-fare")
    public ResponseEntity<ApiResponse<FareCalculationResponseDTO>> calculateFare(@Valid @RequestBody FareCalculationRequestDTO requestDTO) {
        FareCalculationResponseDTO response = paymentService.calculateFare(requestDTO);
        return ResponseEntity.ok(ApiResponse.success("Fare calculated successfully", response));
    }

    @GetMapping("/wallet/{userId}")
    public ResponseEntity<ApiResponse<Wallet>> getWalletByUserId(@PathVariable Integer userId) {
        Wallet wallet = walletService.getWalletByUserId(userId);
        return ResponseEntity.ok(ApiResponse.success("Wallet balance retrieved", wallet));
    }

    @PostMapping("/wallet/add-funds")
    public ResponseEntity<ApiResponse<Wallet>> topUpWallet(@Valid @RequestBody WalletRequestDTO requestDTO) {
        Wallet wallet = walletService.topUpWallet(requestDTO);
        return ResponseEntity.ok(ApiResponse.success("Wallet topped up successfully", wallet));
    }

    @GetMapping("/wallet/{userId}/transactions")
    public ResponseEntity<ApiResponse<List<WalletTransaction>>> getWalletTransactions(@PathVariable Integer userId) {
        List<WalletTransaction> transactions = walletService.getWalletTransactions(userId);
        return ResponseEntity.ok(ApiResponse.success("Wallet transactions retrieved", transactions));
    }
}

File: app/backend/.net/SmartRide.PaymentService/Controllers/PaymentController.cs
using Microsoft.AspNetCore.Mvc;
using SmartRide.PaymentService.Models.DTOs;
using SmartRide.PaymentService.Models.Entities;
using SmartRide.PaymentService.Services;

namespace SmartRide.PaymentService.Controllers;

[ApiController]
[Route("api/payments")]
public class PaymentController : ControllerBase
{
    private readonly IPaymentService _paymentService;
    private readonly IWalletService _walletService;

    public PaymentController(IPaymentService paymentService, IWalletService walletService)
    {
        _paymentService = paymentService;
        _walletService = walletService;
    }

    [HttpGet("health")]
    public ActionResult<ApiResponse<string>> HealthCheck()
    {
        return Ok(ApiResponse<string>.SuccessResponse("Payment Service is UP and running", "OK"));
    }

    [HttpGet]
    public async Task<ActionResult<ApiResponse<List<PaymentResponseDTO>>>> GetAllPayments()
    {
        var response = await _paymentService.GetAllPaymentsAsync();
        return Ok(ApiResponse<List<PaymentResponseDTO>>.SuccessResponse("All payments retrieved", response));
    }

    [HttpGet("all")]
    public async Task<ActionResult<ApiResponse<List<PaymentResponseDTO>>>> GetAllPaymentsAlias()
    {
        var response = await _paymentService.GetAllPaymentsAsync();
        return Ok(ApiResponse<List<PaymentResponseDTO>>.SuccessResponse("All payments retrieved", response));
    }

    [HttpPost("process")]
    public async Task<ActionResult<ApiResponse<PaymentResponseDTO>>> ProcessPayment([FromBody] PaymentRequestDTO requestDTO)
    {
        var response = await _paymentService.ProcessPaymentAsync(requestDTO);
        return StatusCode(201, ApiResponse<PaymentResponseDTO>.SuccessResponse("Payment processed successfully", response));
    }

    [HttpGet("{paymentId:int}")]
    public async Task<ActionResult<ApiResponse<PaymentResponseDTO>>> GetPaymentById(int paymentId)
    {
        var response = await _paymentService.GetPaymentByIdAsync(paymentId);
        return Ok(ApiResponse<PaymentResponseDTO>.SuccessResponse("Payment details retrieved", response));
    }

    [HttpGet("transaction/{transactionId}")]
    public async Task<ActionResult<ApiResponse<PaymentResponseDTO>>> GetPaymentByTransactionId(string transactionId)
    {
        var response = await _paymentService.GetPaymentByTransactionIdAsync(transactionId);
        return Ok(ApiResponse<PaymentResponseDTO>.SuccessResponse("Payment details retrieved", response));
    }

    [HttpGet("ride/{rideId:int}")]
    public async Task<ActionResult<ApiResponse<PaymentResponseDTO>>> GetPaymentByRideId(int rideId)
    {
        var response = await _paymentService.GetPaymentByRideIdAsync(rideId);
        return Ok(ApiResponse<PaymentResponseDTO>.SuccessResponse("Payment details retrieved", response));
    }

    [HttpGet("user/{userId:int}")]
    public async Task<ActionResult<ApiResponse<List<PaymentResponseDTO>>>> GetPaymentsByUserId(int userId)
    {
        var response = await _paymentService.GetPaymentsByUserIdAsync(userId);
        return Ok(ApiResponse<List<PaymentResponseDTO>>.SuccessResponse("User payment history retrieved", response));
    }

    [HttpPost("refund")]
    public async Task<ActionResult<ApiResponse<PaymentResponseDTO>>> RefundPayment([FromBody] RefundRequestDTO requestDTO)
    {
        var response = await _paymentService.RefundPaymentAsync(requestDTO);
        return Ok(ApiResponse<PaymentResponseDTO>.SuccessResponse("Payment refunded successfully", response));
    }

    [HttpPost("calculate-fare")]
    public ActionResult<ApiResponse<FareCalculationResponseDTO>> CalculateFare([FromBody] FareCalculationRequestDTO requestDTO)
    {
        var response = _paymentService.CalculateFare(requestDTO);
        return Ok(ApiResponse<FareCalculationResponseDTO>.SuccessResponse("Fare calculated successfully", response));
    }

    [HttpGet("wallet/{userId:int}")]
    public async Task<ActionResult<ApiResponse<Wallet>>> GetWalletByUserId(int userId)
    {
        var wallet = await _walletService.GetWalletByUserIdAsync(userId);
        return Ok(ApiResponse<Wallet>.SuccessResponse("Wallet balance retrieved", wallet));
    }

    [HttpPost("wallet/add-funds")]
    public async Task<ActionResult<ApiResponse<Wallet>>> TopUpWallet([FromBody] WalletRequestDTO requestDTO)
    {
        var wallet = await _walletService.TopUpWalletAsync(requestDTO);
        return Ok(ApiResponse<Wallet>.SuccessResponse("Wallet topped up successfully", wallet));
    }

    [HttpGet("wallet/{userId:int}/transactions")]
    public async Task<ActionResult<ApiResponse<List<WalletTransaction>>>> GetWalletTransactions(int userId)
    {
        var transactions = await _walletService.GetWalletTransactionsAsync(userId);
        return Ok(ApiResponse<List<WalletTransaction>>.SuccessResponse("Wallet transactions retrieved", transactions));
    }
}
